### PR TITLE
feat: customize retry behavior implementation (#1474)

### PR DESCRIPTION
**Pull request by shaffeeullah**
Retry customization implementation. 

Adds the RetryOptions interface which allows users to configure retryDelayMultiplier, totalTimeout, maxRetryDelay, autoRetry, maxRetries, and retryableErrorFn

### CODE REVIEW

**Review by ddelgrosso1 — APPROVED** [review]
LGTM

File: src/storage.ts [hunk 1]
```diff
@@ -163,6 +177,77 @@ export type GetHmacKeysResponse = [HmacKey[]];
...
+  if (err) {
+    if ([408, 429, 500, 502, 503, 504].indexOf(err.code!) !== -1) {
+      return true;
+    }
+
+    if (err.errors) {
+      for (const e of err.errors) {
+        const reason = e.reason?.toLowerCase();
+        if (
+          (reason && reason.includes('eai_again')) ||
```
Comment (danielduhh) [inline src/storage.ts:239]: What does this `eai_again` mean? Maybe add a comment?

File: src/storage.ts [hunk 2]
```diff
@@ -163,6 +177,77 @@ export type GetHmacKeysResponse = [HmacKey[]];
...
+          reason === 'connection reset by peer' ||
```
Comment (danielduhh) [inline src/storage.ts:240]: Is it same to assume the text here won't change?

Comment (shaffeeullah) [reply src/storage.ts:240]: There's nothing stopping the GCS team from changing it. Similar logic has been in the released version of nodejs-common (https://github.com/googleapis/nodejs-common/blob/513374ca1bce94ba720106ab78c776a0e61c0358/src/util.ts#L549) and I'm not sure of another way to check for this error. This logic was designed to be similar to common.

File: src/storage.ts (hunk 1 above)
Comment (shaffeeullah) [reply src/storage.ts]: It's a DNS lookup error. Added a comment.

**Review by danielduhh — APPROVED** [review]
🚢

**Pull request merged**